File: UnityMcpBridge/Editor/Tools/ManageBuild.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using UnityEngine;
using UnityEditor;
using UnityEditor.Build.Reporting;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using UnityMcpBridge.Editor.Helpers;

namespace UnityMcpBridge.Tools
{
    public static class ManageBuild
    {
        public static JObject HandleCommand(JObject parameters)
        {
            return HandleBuildCommand(parameters);
        }

        public static JObject HandleBuildCommand(JObject parameters)
        {
            try
            {
                string action = parameters["action"]?.ToString();

                return action switch
                {
                    "create_build" => CreateBuild(parameters),
                    "configure_build" => ConfigureBuild(parameters),
                    "list_builds" => ListBuilds(),
                    "get_build_info" => GetBuildInfo(parameters),
                    "build_project" => BuildProject(parameters),
                    "get_build_log" => GetBuildLog(parameters),
                    "validate_build" => ValidateBuild(parameters),
                    _ => new JObject { ["error"] = $"Unknown action: {action}" }
                };
            }
            catch (Exception e)
            {
                return new JObject { ["error"] = e.Message };
            }
        }

        public static JObject HandleAssetBundleCommand(JObject parameters)
        {
            try
            {
                string action = parameters["action"]?.ToString();

                return action switch
                {
                    "create_bundle" => CreateAssetBundle(parameters),
                    "add_assets" => AddAssetsToBundle(parameters),
                    "remove_assets" => RemoveAssetsFromBundle(parameters),
                    "build_bundles" => BuildAssetBundles(parameters),
                    "list_bundles" => ListAssetBundles(),
                    "get_bundle_info" => GetAssetBundleInfo(parameters),
                    "delete_bundle" => DeleteAssetBundle(parameters),
                    "validate_bundles" => ValidateAssetBundles(),
                    _ => new JObject { ["error"] = $"Unknown action: {action}" }
                };
            }
            catch (Exception e)
            {
                return new JObject { ["error"] = e.Message };
            }
        }

        public static JObject HandleBuildPipelineCommand(JObject parameters)
        {
            try
            {
                string action = parameters["action"]?.ToString();

                return action switch
                {
                    "create_pipeline" => CreateBuildPipeline(parameters),
                    "modify_pipeline" => ModifyBuildPipeline(parameters),
                    "run_pipeline" => RunBuildPipeline(parameters),
                    "list_pipelines" => ListBuildPipelines(),
                    "get_pipeline_info" => GetBuildPipelineInfo(parameters),
                    "delete_pipeline" => DeleteBuildPipeline(parameters),
                    "validate_pipeline" => ValidateBuildPipeline(parameters),
                    "get_pipeline_logs" => GetBuildPipelineLogs(parameters),
                    _ => new JObject { ["error"] = $"Unknown action: {action}" }
                };
            }
            catch (Exception e)
            {
                return new JObject { ["error"] = e.Message };
            }
        }

        public static JObject HandleDeploymentCommand(JObject parameters)
        {
            try
            {
                string action = parameters["action"]?.ToString();

                return action switch
                {
                    "deploy_build" => DeployBuild(parameters),
                    "configure_deployment" => ConfigureDeployment(parameters),
                    "list_deployments" => ListDeployments(),
                    "get_deployment_status" => GetDeploymentStatus(parameters),
                    "rollback_deployment" => RollbackDeployment(parameters),
                    "validate_deployment" => ValidateDeployment(parameters),
                    "get_deployment_logs" => GetDeploymentLogs(parameters),
                    _ => new JObject { ["error"] = $"Unknown action: {action}" }
                };
            }
            catch (Exception e)
            {
                return new JObject { ["error"] = e.Message };
            }
        }

        private static JObject CreateBuild(JObject parameters)
        {
            try
            {
                var buildTarget = ParseBuildTarget(parameters["build_target"]?.ToString() ?? "");
                var buildPath = parameters["build_path"]?.ToString() ?? "";
                var scenes = parameters["scenes"] != null ? 
                    parameters["scenes"].ToObject<string[]>() : 
                    EditorBuildSettings.scenes.Where(s => s.enabled).Select(s => s.path).ToArray();

                var buildOptions = BuildOptions.None;
                if (parameters["development_build"]?.ToObject<bool>() ?? false)
                    buildOptions |= BuildOptions.Development;
                if (parameters["script_debugging"]?.ToObject<bool>() ?? false)
                    buildOptions |= BuildOptions.AllowDebugging;

                var buildPlayerOptions = new BuildPlayerOptions
                {
                    scenes = scenes,
                    locationPathName = buildPath,
                    target = buildTarget,
                    options = buildOptions
                };

                var report = BuildPipeline.BuildPlayer(buildPlayerOptions);
                
                return new JObject
                {
                    ["success"] = report.summary.result == BuildResult.Succeeded,
                    ["result"] = report.summary.result.ToString(),
                    ["totalTime"] = report.summary.totalTime.TotalSeconds,
                    ["totalSize"] = report.summary.totalSize,
                    ["outputPath"] = report.summary.outputPath,
                    ["platform"] = report.summary.platform.ToString()
                };
            }
            catch (Exception e)
            {
                return new JObject { ["error"] = e.Message };
            }
        }

        private static JObject ConfigureBuild(JObject parameters)
        {
            try
            {
                // Configure build settings
                if (parameters["company_name"] != null)
                    PlayerSettings.companyName = parameters["company_name"].ToString();
                if (parameters["product_name"] != null)
                    PlayerSettings.productName = parameters["product_name"].ToString();
                if (parameters["version"] != null)
                    PlayerSettings.bundleVersion = parameters["version"].ToString();

                return new JObject { ["success"] = true, ["message"] = "Build settings configured" };
            }
            catch (Exception e)
            {
                return new JObject { ["error"] = e.Message };
            }
        }

        private static JObject ListBuilds()
        {
            try
            {
                var builds = new JArray();
                
                // Get available build targets
                foreach (BuildTarget target in Enum.GetValues(typeof(BuildTarget)))
                {
                    if (BuildPipeline.IsBuildTargetSupported(BuildTargetGroup.Standalone, target))
                    {
                        builds.Add(new JObject
                        {
                            ["target"] = target.ToString(),
                            ["supported"] = true,
                            ["group"] = BuildPipeline.GetBuildTargetGroup(target).ToString()
                        });
                    }
                }

                return new JObject { ["builds"] = builds };
            }
            catch (Exception e)
            {
                return new JObject { ["error"] = e.Message };
            }
        }

        private static JObject GetBuildInfo(JObject parameters)
        {
            try
            {
                var buildTarget = ParseBuildTarget(parameters["build_target"]?.ToString() ?? "");
                
                var info = new JObject
                {
                    ["target"] = buildTarget.ToString(),
                    ["group"] = BuildPipeline.GetBuildTargetGroup(buildTarget).ToString(),
                    ["supported"] = BuildPipeline.IsBuildTargetSupported(BuildPipeline.GetBuildTargetGroup(buildTarget), buildTarget),
                    ["scenes"] = new JArray(EditorBuildSettings.scenes.Select(s => new JObject { ["path"] = s.path, ["enabled"] = s.enabled })),
                    ["playerSettings"] = new JObject
                    {
                        ["companyName"] = PlayerSettings.companyName,
                        ["productName"] = PlayerSettings.productName,
                        ["version"] = PlayerSettings.bundleVersion
                    }
                };

                return info;
            }
            catch (Exception e)
            {
                return new JObject { ["error"] = e.Message };
            }
        }

        private static JObject BuildProject(JObject parameters)
        {
            return CreateBuild(parameters); // Same as create_build
        }

        private static JObject GetBuildLog(JObject parameters)
        {
            try
            {
                // Placeholder for build log retrieval
                return new JObject 
                { 
                    ["message"] = "Build log retrieval not yet implemented",
                    ["logs"] = new JArray("Build log functionality requires custom implementation")
                };
            }
            catch (Exception e)
            {
                return new JObject { ["error"] = e.Message };
            }
        }

        private static JObject ValidateBuild(JObject parameters)
        {
            try
            {
                var issues = new JArray();
                
                // Check if scenes are set
                if (EditorBuildSettings.scenes.Length == 0)
                    issues.Add("No scenes configured for build");
                
                // Check if build target is supported
                var buildTarget = ParseBuildTarget(parameters["build_target"]?.ToString() ?? "");
                if (!BuildPipeline.IsBuildTargetSupported(BuildPipeline.GetBuildTargetGroup(buildTarget), buildTarget))
                    issues.Add($"Build target {buildTarget} is not supported");

                return new JObject 
                { 
                    ["valid"] = issues.Count == 0,
                    ["issues"] = issues
                };
            }
            catch (Exception e)
            {
                return new JObject { ["error"] = e.Message };
            }
        }

        // Asset Bundle Methods
        private static string CreateAssetBundle(Dictionary<string, object> data)
        {
            try
            {
                var bundleName = data.GetValueOrDefault("bundle_name", "").ToString();
                var assets = data.ContainsKey("assets") ? 
                    JsonConvert.DeserializeObject<string[]>(data["assets"].ToString()) : new string[0];

                foreach (var assetPath in assets)
                {
                    var importer = AssetImporter.GetAtPath(assetPath);
                    if (importer != null)
                    {
                        importer.assetBundleName = bundleName;
                    }
                }

                return JsonConvert.SerializeObject(new 
                { 
                    success = true, 
                    bundleName = bundleName,
                    assetsCount = assets.Length
                });
            }
            catch (Exception e)
            {
                return JsonConvert.SerializeObject(new { error = e.Message });
            }
        }

        private static string AddAssetsToBundle(Dictionary<string, object> data)
        {
            return CreateAssetBundle(data); // Same logic
        }

        private static string RemoveAssetsFromBundle(Dictionary<string, object> data)
        {
            try
            {
                var assets = data.ContainsKey("assets") ? 
                    JsonConvert.DeserializeObject<string[]>(data["assets"].ToString()) : new string[0];

                foreach (var assetPath in assets)
                {
                    var importer = AssetImporter.GetAtPath(assetPath);
                    if (importer != null)
                    {
                        importer.assetBundleName = "";
                    }
                }

                return JsonConvert.SerializeObject(new 
                { 
                    success = true,
                    removedAssetsCount = assets.Length
                });
            }
            catch (Exception e)
            {
                return JsonConvert.SerializeObject(new { error = e.Message });
            }
        }

        private static string BuildAssetBundles(Dictionary<string, object> data)
        {
            try
            {
                var buildPath = data.GetValueOrDefault("build_path", "Assets/StreamingAssets").ToString();
                var buildTarget = ParseBuildTarget(data.GetValueOrDefault("build_target", EditorUserBuildSettings.activeBuildTarget.ToString()).ToString());

                if (!Directory.Exists(buildPath))
                    Directory.CreateDirectory(buildPath);

                var manifest = BuildPipeline.BuildAssetBundles(buildPath, BuildAssetBundleOptions.None, buildTarget);

                return JsonConvert.SerializeObject(new 
                { 
                    success = manifest != null,
                    buildPath = buildPath,
                    bundles = manifest?.GetAllAssetBundles() ?? new string[0]
                });
            }
            catch (Exception e)
            {
                return JsonConvert.SerializeObject(new { error = e.Message });
            }
        }

        private static string ListAssetBundles()
        {
            try
            {
                var bundleNames = AssetDatabase.GetAllAssetBundleNames();
                var bundles = bundleNames.Select(name => new
                {
                    name = name,
                    assets = AssetDatabase.GetAssetPathsFromAssetBundle(name)
                }).ToArray();

                return JsonConvert.SerializeObject(new { bundles });
            }
            catch (Exception e)
            {
                return JsonConvert.SerializeObject(new { error = e.Message });
            }
        }

        private static string GetAssetBundleInfo(Dictionary<string, object> data)
        {
            try
            {
                var bundleName = data.GetValueOrDefault("bundle_name", "").ToString();
                var assets = AssetDatabase.GetAssetPathsFromAssetBundle(bundleName);

                return JsonConvert.SerializeObject(new 
                { 
                    bundleName = bundleName,
                    assets = assets,
                    assetCount = assets.Length
                });
            }
            catch (Exception e)
            {
                return JsonConvert.SerializeObject(new { error = e.Message });
            }
        }

        private static string DeleteAssetBundle(Dictionary<string, object> data)
        {
            try
            {
                var bundleName = data.GetValueOrDefault("bundle_name", "").ToString();
                AssetDatabase.RemoveAssetBundleName(bundleName, true);

                return JsonConvert.SerializeObject(new 
                { 
                    success = true,
                    message = $"Asset bundle '{bundleName}' deleted"
                });
            }
            catch (Exception e)
            {
                return JsonConvert.SerializeObject(new { error = e.Message });
            }
        }

        private static string ValidateAssetBundles()
        {
            try
            {
                var issues = new List<string>();
                var bundleNames = AssetDatabase.GetAllAssetBundleNames();

                foreach (var bundleName in bundleNames)
                {
                    var assets = AssetDatabase.GetAssetPathsFromAssetBundle(bundleName);
                    if (assets.Length == 0)
                        issues.Add($"Bundle '{bundleName}' has no assets");
                }

                return JsonConvert.SerializeObject(new 
                { 
                    valid = issues.Count == 0,
                    issues = issues.ToArray(),
                    bundleCount = bundleNames.Length
                });
            }
            catch (Exception e)
            {
                return JsonConvert.SerializeObject(new { error = e.Message });
            }
        }

        // Placeholder methods for build pipeline and deployment
        private static string CreateBuildPipeline(Dictionary<string, object> data)
        {
            return JsonConvert.SerializeObject(new { message = "Build pipeline creation not yet implemented" });
        }

        private static string ModifyBuildPipeline(Dictionary<string, object> data)
        {
            return JsonConvert.SerializeObject(new { message = "Build pipeline modification not yet implemented" });
        }

        private static string RunBuildPipeline(Dictionary<string, object> data)
        {
            return JsonConvert.SerializeObject(new { message = "Build pipeline execution not yet implemented" });
        }

        private static string ListBuildPipelines()
        {
            return JsonConvert.SerializeObject(new { message = "Build pipeline listing not yet implemented" });
        }

        private static string GetBuildPipelineInfo(Dictionary<string, object> data)
        {
            return JsonConvert.SerializeObject(new { message = "Build pipeline info not yet implemented" });
        }

        private static string DeleteBuildPipeline(Dictionary<string, object> data)
        {
            return JsonConvert.SerializeObject(new { message = "Build pipeline deletion not yet implemented" });
        }

        private static string ValidateBuildPipeline(Dictionary<string, object> data)
        {
            return JsonConvert.SerializeObject(new { message = "Build pipeline validation not yet implemented" });
        }

        private static string GetBuildPipelineLogs(Dictionary<string, object> data)
        {
            return JsonConvert.SerializeObject(new { message = "Build pipeline logs not yet implemented" });
        }

        private static string DeployBuild(Dictionary<string, object> data)
        {
            return JsonConvert.SerializeObject(new { message = "Build deployment not yet implemented" });
        }

        private static string ConfigureDeployment(Dictionary<string, object> data)
        {
            return JsonConvert.SerializeObject(new { message = "Deployment configuration not yet implemented" });
        }

        private static string ListDeployments()
        {
            return JsonConvert.SerializeObject(new { message = "Deployment listing not yet implemented" });
        }

        private static string GetDeploymentStatus(Dictionary<string, object> data)
        {
            return JsonConvert.SerializeObject(new { message = "Deployment status not yet implemented" });
        }

        private static string RollbackDeployment(Dictionary<string, object> data)
        {
            return JsonConvert.SerializeObject(new { message = "Deployment rollback not yet implemented" });
        }

        private static string ValidateDeployment(Dictionary<string, object> data)
        {
            return JsonConvert.SerializeObject(new { message = "Deployment validation not yet implemented" });
        }

        private static string GetDeploymentLogs(Dictionary<string, object> data)
        {
            return JsonConvert.SerializeObject(new { message = "Deployment logs not yet implemented" });
        }

        private static BuildTarget ParseBuildTarget(string targetString)
        {
            if (Enum.TryParse<BuildTarget>(targetString, true, out var target))
                return target;
            return EditorUserBuildSettings.activeBuildTarget;
        }
    }
}

File: UnityMcpServer/src/tools/manage_build.py
"""
Unity Build and Deployment Management Tool

This tool provides comprehensive build and deployment management for Unity projects,
including build configuration, platform-specific builds, asset bundling, and deployment automation.
"""

import json
from typing import Dict, List, Any, Optional
from mcp import types
from mcp.server import Server

def register_build_tools(server: Server):
    """Register all build and deployment management tools"""
    
    @server.tool()
    async def manage_build(
        action: str = None,
        build_target: Optional[str] = None,
        build_path: Optional[str] = None,
        build_options: Optional[Dict[str, Any]] = None,
        scenes: Optional[List[str]] = None,
        development_build: Optional[bool] = None,
        script_debugging: Optional[bool] = None,
        compression: Optional[str] = None,
        **kwargs
    ) -> List[types.TextContent]:
        """
        Manage Unity build operations and configurations.
        
        Actions:
        - create_build: Create a new build with specified settings
        - configure_build: Configure build settings and options
        - list_builds: List available build configurations
        - get_build_info: Get information about a specific build
        - delete_build: Delete a build configuration
        - build_project: Execute the build process
        - get_build_log: Get build log information
        - validate_build: Validate build settings and requirements
        
        Args:
            action: The build action to perform
            build_target: Target platform (Standalone, Android, iOS, WebGL, etc.)
            build_path: Output path for the build
            build_options: Additional build options and settings
            scenes: List of scenes to include in the build
            development_build: Whether to create a development build
            script_debugging: Enable script debugging
            compression: Compression method for the build
        """
        
        try:
            # Prepare command data
            command_data = {
                "action": action,
                "build_target": build_target,
                "build_path": build_path,
                "build_options": build_options or {},
                "scenes": scenes or [],
                "development_build": development_build,
                "script_debugging": script_debugging,
                "compression": compression,
                **kwargs
            }
            
            # Send command to Unity
            from ..core.unity_bridge import send_command_to_unity
            # Unity bridge expects 'HandleManageBuild' as the type in the registry
            result = await send_command_to_unity("HandleManageBuild", command_data)
            
            return [types.TextContent(
                type="text",
                text=f"Build management result: {json.dumps(result, indent=2)}"
            )]
            
        except Exception as e:
            return [types.TextContent(
                type="text", 
                text=f"Error in build management: {str(e)}"
            )]
    
    @server.tool()
    async def asset_bundle_operations(
        action: str,
        bundle_name: Optional[str] = None,
        assets: Optional[List[str]] = None,
        build_path: Optional[str] = None,
        build_target: Optional[str] = None,
        compression: Optional[str] = None,
        **kwargs
    ) -> List[types.TextContent]:
        """
        Manage Unity Asset Bundle operations.
        
        Actions:
        - create_bundle: Create a new asset bundle
        - add_assets: Add assets to an existing bundle
        - remove_assets: Remove assets from a bundle
        - build_bundles: Build all asset bundles
        - list_bundles: List all asset bundles
        - get_bundle_info: Get information about a specific bundle
        - delete_bundle: Delete an asset bundle
        - validate_bundles: Validate asset bundle dependencies
        
        Args:
            action: The asset bundle action to perform
            bundle_name: Name of the asset bundle
            assets: List of asset paths to include
            build_path: Output path for built bundles
            build_target: Target platform for bundles
            compression: Compression method for bundles
        """
        
        try:
            command_data = {
                "action": action,
                "bundle_name": bundle_name,
                "assets": assets or [],
                "build_path": build_path,
                "build_target": build_target,
                "compression": compression,
                **kwargs
            }
            
            from ..core.unity_bridge import send_command_to_unity
            result = await send_command_to_unity("HandleManageBuild", command_data)
            
            return [types.TextContent(
                type="text",
                text=f"Asset bundle operation result: {json.dumps(result, indent=2)}"
            )]
            
        except Exception as e:
            return [types.TextContent(
                type="text",
                text=f"Error in asset bundle operations: {str(e)}"
            )]
    
    @server.tool()
    async def build_pipeline_operations(
        action: str,
        pipeline_name: Optional[str] = None,
        steps: Optional[List[Dict[str, Any]]] = None,
        triggers: Optional[List[str]] = None,
        environment: Optional[Dict[str, str]] = None,
        **kwargs
    ) -> List[types.TextContent]:
        """
        Manage Unity build pipeline and automation.
        
        Actions:
        - create_pipeline: Create a new build pipeline
        - modify_pipeline: Modify an existing pipeline
        - run_pipeline: Execute a build pipeline
        - list_pipelines: List all build pipelines
        - get_pipeline_info: Get information about a pipeline
        - delete_pipeline: Delete a build pipeline
        - validate_pipeline: Validate pipeline configuration
        - get_pipeline_logs: Get pipeline execution logs
        
        Args:
            action: The pipeline action to perform
            pipeline_name: Name of the build pipeline
            steps: List of pipeline steps and configurations
            triggers: List of pipeline triggers (manual, git, schedule)
            environment: Environment variables for the pipeline
        """
        
        try:
            command_data = {
                "action": action,
                "pipeline_name": pipeline_name,
                "steps": steps or [],
                "triggers": triggers or [],
                "environment": environment or {},
                **kwargs
            }
            
            from ..core.unity_bridge import send_command_to_unity
            result = await send_command_to_unity("HandleManageBuild", command_data)
            
            return [types.TextContent(
                type="text",
                text=f"Build pipeline operation result: {json.dumps(result, indent=2)}"
            )]
            
        except Exception as e:
            return [types.TextContent(
                type="text",
                text=f"Error in build pipeline operations: {str(e)}"
            )]
    
    @server.tool()
    async def deployment_operations(
        action: str,
        deployment_target: Optional[str] = None,
        build_path: Optional[str] = None,
        credentials: Optional[Dict[str, str]] = None,
        deployment_config: Optional[Dict[str, Any]] = None,
        **kwargs
    ) -> List[types.TextContent]:
        """
        Manage Unity project deployment operations.
        
        Actions:
        - deploy_build: Deploy a build to specified target
        - configure_deployment: Configure deployment settings
        - list_deployments: List deployment configurations
        - get_deployment_status: Get status of a deployment
        - rollback_deployment: Rollback to previous deployment
        - validate_deployment: Validate deployment configuration
        - get_deployment_logs: Get deployment logs
        
        Args:
            action: The deployment action to perform
            deployment_target: Target platform/service (Steam, App Store, Google Play, etc.)
            build_path: Path to the build to deploy
            credentials: Deployment credentials and authentication
            deployment_config: Deployment-specific configuration
        """
        
        try:
            command_data = {
                "action": action,
                "deployment_target": deployment_target,
                "build_path": build_path,
                "credentials": credentials or {},
                "deployment_config": deployment_config or {},
                **kwargs
            }
            
            from ..core.unity_bridge import send_command_to_unity
            result = await send_command_to_unity("HandleManageBuild", command_data)
            
            return [types.TextContent(
                type="text",
                text=f"Deployment operation result: {json.dumps(result, indent=2)}"
            )]
            
        except Exception as e:
            return [types.TextContent(
                type="text",
                text=f"Error in deployment operations: {str(e)}"
            )]
    
    return manage_build